File: src/lidar_icp_node.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud2
import sensor_msgs.point_cloud2 as pc2
import pcl
import tf
import tf2_ros
import numpy as np
from std_msgs.msg import Header
from geometry_msgs.msg import TransformStamped

class LidarICPNode:
    def __init__(self):
        rospy.init_node('LidarICPNode', anonymous=True)
        
        # ICP (Iterative Closest Point) Parameters
        self.max_correspondence_distance = 0.5  # Maximum distance (in meters) between points to be considered a match
        self.max_iterations = 50                # Number of ICP iterations before giving up
        self.fitness_epsilon = 1e-6             # Stop iterating if improvement is less than this value

        # Point Cloud Processing Parameters
        self.leaf_size = 0.2                    # Size (in meters) of voxel grid cells for downsampling point cloud
        self.min_fitness_score = 0.01           # Maximum acceptable average distance between matched points (lower is better)

        # Motion Limit Parameters
        self.max_translation = 5.0              # Maximum allowed movement (in meters) between consecutive frames
        self.max_rotation = 1.0                 # Maximum allowed rotation (in radians) between consecutive frames (~57 degrees)
        
        # Subscribers and Publishers
        self.subscriber = rospy.Subscriber('/os_cloud_node/points', PointCloud2, self.callback)
        self.pub_aligned = rospy.Publisher('/aligned_points', PointCloud2, queue_size=1)
        self.pub_current = rospy.Publisher('/current_aligned', PointCloud2, queue_size=1)
        
        # TF handling
        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        self.fixed_frame = "map"  # Changed from t0_static_frame to map
        
        # State variables
        self.previous_cloud = None
        self.cumulative_transform = np.identity(4)
        self.accumulated_points = []
        self.callback_count = 0

    def filter_cloud(self, cloud_pcl):
        """Apply filtering to input point cloud."""
        # Voxel Grid filter for downsampling
        vg = cloud_pcl.make_voxel_grid_filter()
        vg.set_leaf_size(self.leaf_size, self.leaf_size, self.leaf_size)
        filtered = vg.filter()
        
        # Statistical Outlier Removal with adjusted parameters
        sor = filtered.make_statistical_outlier_filter()
        sor.set_mean_k(30)  # Reduced from 50
        sor.set_std_dev_mul_thresh(2.0)  # Increased from 1.0
        filtered = sor.filter()
        
        return filtered

    def check_transform_validity(self, transform):
        """Check if the transformation is within acceptable limits."""
        # Check translation magnitude
        translation = transform[:3, 3]
        translation_magnitude = np.linalg.norm(translation)
        if translation_magnitude > self.max_translation:
            return False, "Excessive translation detected"
        
        # Check rotation magnitude
        rotation = transform[:3, :3]
        euler_angles = tf.transformations.euler_from_matrix(rotation)
        max_rotation = max(abs(angle) for angle in euler_angles)
        if max_rotation > self.max_rotation:
            return False, "Excessive rotation detected"
        
        return True, "Transform within limits"

    def perform_icp(self, source_cloud, target_cloud):
        """Perform ICP with error checking."""
        # Create ICP object
        icp = source_cloud.make_IterativeClosestPoint()
        
        # Perform ICP
        converged, transf, estimate, fitness = icp.icp(source_cloud, target_cloud)
        
        rospy.loginfo(f"ICP Fitness Score: {fitness}")
        rospy.loginfo(f"Convergence Status: {converged}")
        
        if not converged:
            return False, None, "ICP failed to converge"
        
        # Only check fitness if convergence succeeded
        if fitness < self.min_fitness_score:
            return False, None, f"Poor alignment score: {fitness}"
        
        # Check if transform is reasonable
        valid_transform, message = self.check_transform_validity(transf)
        if not valid_transform:
            # Log the actual translation for debugging
            translation = transf[:3, 3]
            translation_magnitude = np.linalg.norm(translation)
            rospy.loginfo(f"Translation magnitude: {translation_magnitude}")
            return False, None, message
        
        return True, transf, f"Success with fitness: {fitness}"

    def transform_cloud(self, cloud_pcl, transform):
        """Transform point cloud using transformation matrix."""
        points = np.asarray(cloud_pcl)
        transformed_points = []
        
        for point in points:
            p = np.array([point[0], point[1], point[2], 1.0])
            transformed = np.dot(transform, p)
            transformed_points.append(transformed[:3])
        
        transformed_cloud = pcl.PointCloud()
        transformed_cloud.from_array(np.array(transformed_points, dtype=np.float32))
        return transformed_cloud

    def publish_transform(self, transform, timestamp, child_frame):
        """Publish transform to TF tree."""
        t = TransformStamped()
        t.header.stamp = timestamp
        t.header.frame_id = self.fixed_frame
        t.child_frame_id = child_frame
        
        t.transform.translation.x = transform[0, 3]
        t.transform.translation.y = transform[1, 3]
        t.transform.translation.z = transform[2, 3]
        
        quat = tf.transformations.quaternion_from_matrix(transform)
        t.transform.rotation.x = quat[0]
        t.transform.rotation.y = quat[1]
        t.transform.rotation.z = quat[2]
        t.transform.rotation.w = quat[3]
        
        self.tf_broadcaster.sendTransform(t)

    def callback(self, data):
        """Main callback for processing incoming point clouds."""
        self.callback_count += 1
        rospy.loginfo(f"\n------ Processing frame {self.callback_count} ------")
        
        # Convert ROS message to PCL
        points_list = []
        for point in pc2.read_points(data, field_names=("x", "y", "z"), skip_nans=True):
            points_list.append([point[0], point[1], point[2]])
        
        if not points_list:
            rospy.logwarn("Empty point cloud received")
            return
        
        current_cloud = pcl.PointCloud()
        current_cloud.from_list(points_list)
        
        # Filter current cloud
        current_filtered = self.filter_cloud(current_cloud)
        rospy.loginfo(f"Filtered cloud size: {current_filtered.size}")
        
        # Handle first frame
        if self.previous_cloud is None:
            self.previous_cloud = current_filtered
            self.accumulated_points = points_list
            
            # Publish initial cloud in fixed frame
            header = Header()
            header.stamp = data.header.stamp
            header.frame_id = self.fixed_frame
            cloud_msg = pc2.create_cloud_xyz32(header, points_list)
            self.pub_aligned.publish(cloud_msg)
            
            # Publish initial transform
            self.publish_transform(self.cumulative_transform, data.header.stamp, data.header.frame_id)
            return
        
        # Perform ICP with previous frame
        success, transform, message = self.perform_icp(current_filtered, self.previous_cloud)
        
        if not success:
            rospy.logwarn(f"ICP failed: {message}")
            return
        
        # Update transforms and points
        self.cumulative_transform = np.dot(self.cumulative_transform, transform)
        
        # Transform current points to fixed frame
        transformed_cloud = self.transform_cloud(current_filtered, self.cumulative_transform)
        transformed_points = transformed_cloud.to_array().tolist()
        
        # Add to accumulated points (with optional downsampling)
        self.accumulated_points.extend(transformed_points)
        
        # Publish results
        header = Header()
        header.stamp = data.header.stamp
        header.frame_id = self.fixed_frame
        
        # Publish accumulated cloud
        accumulated_msg = pc2.create_cloud_xyz32(header, self.accumulated_points)
        self.pub_aligned.publish(accumulated_msg)
        
        # Publish current aligned cloud
        current_msg = pc2.create_cloud_xyz32(header, transformed_points)
        self.pub_current.publish(current_msg)
        
        # Publish transform
        self.publish_transform(self.cumulative_transform, data.header.stamp, data.header.frame_id)
        
        # Update previous cloud for next iteration
        self.previous_cloud = current_filtered
        
        rospy.loginfo(f"Successfully processed frame {self.callback_count}")
        rospy.loginfo(f"Total accumulated points: {len(self.accumulated_points)}")

if __name__ == '__main__':
    try:
        node = LidarICPNode()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: src/ros2/src/lidar_icp_node.cpp
#include "lidar_icp/lidar_icp_node.hpp"
#include <pcl_conversions/pcl_conversions.h>

LidarICPNode::LidarICPNode() : Node("lidar_icp_node"), callback_count_(0) {
    // Initialize parameters
    declare_parameter("max_correspondence_distance", 0.5);
    declare_parameter("max_iterations", 50);
    declare_parameter("fitness_epsilon", 1e-6);
    declare_parameter("leaf_size", 0.2);
    declare_parameter("min_fitness_score", 5.0);
    declare_parameter("max_translation", 5.0);
    declare_parameter("max_rotation", 3.15);
    declare_parameter("fixed_frame", "map");

    // Get parameters
    max_correspondence_distance_ = get_parameter("max_correspondence_distance").as_double();
    max_iterations_ = get_parameter("max_iterations").as_int();
    fitness_epsilon_ = get_parameter("fitness_epsilon").as_double();
    leaf_size_ = get_parameter("leaf_size").as_double();
    min_fitness_score_ = get_parameter("min_fitness_score").as_double();
    max_translation_ = get_parameter("max_translation").as_double();
    max_rotation_ = get_parameter("max_rotation").as_double();
    fixed_frame_ = get_parameter("fixed_frame").as_string();

    // Initialize subscribers and publishers
    subscription_ = create_subscription<sensor_msgs::msg::PointCloud2>(
        "points", 10, std::bind(&LidarICPNode::pointCloudCallback, this, std::placeholders::_1));
    
    // Initialize publishers with correct topics
    pub_cloud_ = create_publisher<sensor_msgs::msg::PointCloud2>("accumulated_points", 10);
    pub_current_ = create_publisher<sensor_msgs::msg::PointCloud2>("current_aligned", 10);

    // Initialize TF broadcaster
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);

    // Initialize point clouds
    reference_cloud_ = pcl::PointCloud<pcl::PointXYZ>::Ptr(new pcl::PointCloud<pcl::PointXYZ>);
    accumulated_cloud_ = pcl::PointCloud<pcl::PointXYZ>::Ptr(new pcl::PointCloud<pcl::PointXYZ>);
    reference_to_map_ = Eigen::Matrix4f::Identity();
}

pcl::PointCloud<pcl::PointXYZ>::Ptr LidarICPNode::transformCloud(
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud,
    const Eigen::Matrix4f& transform) {
    pcl::PointCloud<pcl::PointXYZ>::Ptr transformed_cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::transformPointCloud(*cloud, *transformed_cloud, transform);
    return transformed_cloud;
}

std::tuple<bool, Eigen::Matrix4f, std::string> LidarICPNode::performICP(
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& source,
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& target) {
    
    pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
    
    // More robust ICP settings
    icp.setMaxCorrespondenceDistance(max_correspondence_distance_);
    icp.setMaximumIterations(max_iterations_);
    icp.setTransformationEpsilon(fitness_epsilon_);
    icp.setEuclideanFitnessEpsilon(0.01);
    icp.setRANSACOutlierRejectionThreshold(0.05);
    icp.setInputSource(source);
    icp.setInputTarget(target);
    
    pcl::PointCloud<pcl::PointXYZ> aligned;
    icp.align(aligned);
    
    double fitness_score = icp.getFitnessScore();
    
    // Detailed logging
    RCLCPP_INFO(get_logger(), "ICP Details:");
    RCLCPP_INFO(get_logger(), "  Converged: %s", icp.hasConverged() ? "true" : "false");
    RCLCPP_INFO(get_logger(), "  Fitness Score: %f", fitness_score);
    
    // Get transformation and log its details
    Eigen::Matrix4f transformation = icp.getFinalTransformation();
    Eigen::Vector3f translation = transformation.block<3, 1>(0, 3);
    Eigen::Vector3f euler = transformation.block<3, 3>(0, 0).eulerAngles(0, 1, 2);
    
    RCLCPP_INFO(get_logger(), "Transform Details:");
    RCLCPP_INFO(get_logger(), "  Translation: [%.3f, %.3f, %.3f] (norm: %.3f)",
                translation.x(), translation.y(), translation.z(), translation.norm());
    RCLCPP_INFO(get_logger(), "  Rotation (rad): [%.3f, %.3f, %.3f] (max: %.3f)",
                euler.x(), euler.y(), euler.z(), euler.array().abs().maxCoeff());

    if (!icp.hasConverged()) {
        return std::make_tuple(false, Eigen::Matrix4f::Identity(), "ICP failed to converge");
    }
    
    if (fitness_score > min_fitness_score_) {
        return std::make_tuple(false, Eigen::Matrix4f::Identity(), 
                             "Poor alignment score: " + std::to_string(fitness_score));
    }
    
    if (!checkTransformValidity(transformation)) {
        return std::make_tuple(false, Eigen::Matrix4f::Identity(), "Invalid transform detected");
    }
    
    return std::make_tuple(true, transformation, 
                          "Success with fitness: " + std::to_string(fitness_score));
}

pcl::PointCloud<pcl::PointXYZ>::Ptr LidarICPNode::filterCloud(
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud) {
    
    auto filtered = pcl::PointCloud<pcl::PointXYZ>::Ptr(new pcl::PointCloud<pcl::PointXYZ>);
    
    // Voxel grid filter with smaller leaf size for more detail
    pcl::VoxelGrid<pcl::PointXYZ> vg;
    vg.setInputCloud(cloud);
    vg.setLeafSize(leaf_size_, leaf_size_, leaf_size_);
    vg.filter(*filtered);
    
    // More aggressive outlier removal
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor;
    sor.setInputCloud(filtered);
    sor.setMeanK(50);  // Increased from 30
    sor.setStddevMulThresh(1.0);  // Decreased from 2.0 for stricter filtering
    sor.filter(*filtered);
    
    return filtered;
}

void LidarICPNode::publishTransform(
    const Eigen::Matrix4f& transform,
    const rclcpp::Time& timestamp,
    const std::string& child_frame) {
    
    geometry_msgs::msg::TransformStamped t;
    t.header.stamp = timestamp;
    t.header.frame_id = fixed_frame_;
    t.child_frame_id = child_frame;
    
    t.transform.translation.x = transform(0, 3);
    t.transform.translation.y = transform(1, 3);
    t.transform.translation.z = transform(2, 3);
    
    Eigen::Quaternionf q(transform.block<3, 3>(0, 0));
    t.transform.rotation.x = q.x();
    t.transform.rotation.y = q.y();
    t.transform.rotation.z = q.z();
    t.transform.rotation.w = q.w();
    
    tf_broadcaster_->sendTransform(t);
}

void LidarICPNode::pointCloudCallback(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
    callback_count_++;
    RCLCPP_INFO(get_logger(), "\n------ Processing frame %d ------", callback_count_);
    
    // Convert ROS message to PCL cloud
    pcl::PointCloud<pcl::PointXYZ>::Ptr current_cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::fromROSMsg(*msg, *current_cloud);
    
    if (current_cloud->empty()) {
        RCLCPP_WARN(get_logger(), "Empty point cloud received");
        return;
    }
    
    // Filter current cloud
    auto current_filtered = filterCloud(current_cloud);
    RCLCPP_INFO(get_logger(), "Filtered cloud size: %lu", current_filtered->size());
    
    // Handle first frame
    if (reference_cloud_->empty()) {
        *reference_cloud_ = *current_filtered;
        *accumulated_cloud_ = *current_filtered;  // Initialize accumulated cloud
        
        // Publish initial clouds and transform
        sensor_msgs::msg::PointCloud2 cloud_msg;
        pcl::toROSMsg(*accumulated_cloud_, cloud_msg);
        cloud_msg.header = msg->header;
        cloud_msg.header.frame_id = fixed_frame_;
        
        pub_cloud_->publish(cloud_msg);  // Publish accumulated
        pub_current_->publish(cloud_msg); // Publish current
        publishTransform(reference_to_map_, msg->header.stamp, msg->header.frame_id);
        return;
    }
    
    // Perform ICP with reference frame
    auto [success, transform, message] = performICP(current_filtered, reference_cloud_);
    
    if (!success) {
        RCLCPP_WARN(get_logger(), "ICP failed: %s", message.c_str());
        return;
    }
    
    // Update cumulative transform and transform current cloud
    reference_to_map_ = reference_to_map_ * transform.inverse();  // Update cumulative transform
    auto transformed_cloud = transformCloud(current_filtered, reference_to_map_);
    
    // Add transformed points to accumulated cloud
    *accumulated_cloud_ += *transformed_cloud;
    
    // Optional: Downsample accumulated cloud if it gets too large
    if (accumulated_cloud_->size() > 100000) {  // Adjust threshold as needed
        accumulated_cloud_ = filterCloud(accumulated_cloud_);
    }
    
    // Publish accumulated cloud
    sensor_msgs::msg::PointCloud2 accumulated_msg;
    pcl::toROSMsg(*accumulated_cloud_, accumulated_msg);
    accumulated_msg.header = msg->header;
    accumulated_msg.header.frame_id = fixed_frame_;
    pub_cloud_->publish(accumulated_msg);
    
    // Publish current aligned cloud
    sensor_msgs::msg::PointCloud2 current_msg;
    pcl::toROSMsg(*transformed_cloud, current_msg);
    current_msg.header = msg->header;
    current_msg.header.frame_id = fixed_frame_;
    pub_current_->publish(current_msg);
    
    // Publish transform
    publishTransform(reference_to_map_, this->now(), msg->header.frame_id);
    
    // Update reference cloud for next iteration
    *reference_cloud_ = *current_filtered;
    
    RCLCPP_INFO(get_logger(), "Successfully processed frame %d", callback_count_);
    RCLCPP_INFO(get_logger(), "Total accumulated points: %lu", accumulated_cloud_->size());
}

bool LidarICPNode::checkTransformValidity(const Eigen::Matrix4f& transform) {
    Eigen::Vector3f translation = transform.block<3, 1>(0, 3);
    if (translation.norm() > max_translation_) {
        RCLCPP_WARN(get_logger(), "Excessive translation detected: %f", translation.norm());
        return false;
    }
    
    Eigen::Vector3f euler = transform.block<3, 3>(0, 0).eulerAngles(0, 1, 2);
    if (euler.array().abs().maxCoeff() > max_rotation_) {
        RCLCPP_WARN(get_logger(), "Excessive rotation detected: %f", euler.array().abs().maxCoeff());
        return false;
    }
    
    return true;
}
